scripts/Package.hs

```diff
@@ -0,0 +1,133 @@
+    callProcess programPath $ ["a", "-y", outputArchivePath] <> packedPaths
+
+buildWithMsBuild :: FilePath -> IO ()
+buildWithMsBuild solutionPath = do
+    let msbuildPath = "C:\\Program Files (x86)\\Microsoft Visual Studio\\2017\\Community\\MSBuild\\15.0\\Bin\\amd64\\MSBuild.exe"
+    callProcess msbuildPath ["/property:Configuration=Release", solutionPath]
+
+copyToDir :: FilePath -> FilePath -> IO ()
+copyToDir destDir sourcePath = do
+    createDirectoryIfMissing True destDir
+    putStrLn $ "Copy " ++ sourcePath ++ " to " ++ destDir
+    let destPath = destDir </> takeFileName sourcePath
+    copyFile sourcePath destPath
+
+pushArtifact :: FilePath -> IO ()
+pushArtifact path = do
+    callProcess "appveyor" ["PushArtifact", path]
+
+-- Function downloads 7z to temp folder, so it doesn't leave any trash behind.
+downloadAndUnpack7z :: FilePath -> FilePath -> IO ()
+downloadAndUnpack7z archiveUrl targetDirectory = do
+    withSystemTempDirectory "" $ \tmpDir -> do
+        let archiveLocalPath = tmpDir </> takeFileName archiveUrl
+        download archiveUrl archiveLocalPath
+        unpack7z archiveLocalPath targetDirectory
+
+-- We need to extract the package with dev libraries and set the environment
+-- variable DATAFRAMES_DEPS_DIR so the MSBuild project recognizes it.
+--
+-- The package contains all dependencies except for Python (with numpy).
+-- Python needs to be provided by CI environment and pointed to by `PythonDir`
+-- environemt variable.
+prepareEnvironment :: FilePath -> IO ()
+prepareEnvironment tempDir = do
+    let depsDirLocal = tempDir </> "deps"
+    downloadAndUnpack7z depsArchiveUrl depsDirLocal
+    setEnv "DATAFRAMES_DEPS_DIR" depsDirLocal
+
+-- Copies subdirectory with all its contents between two directories
+copyDirectory :: FilePath -> FilePath -> FilePath -> IO ()
+copyDirectory sourceDirectory targetDirectory subdirectoryFilename = do
+    let from = sourceDirectory </> subdirectoryFilename
+    let to = targetDirectory </> subdirectoryFilename
+    copyDirectoryRecursive silent from to
+
+main :: IO ()
+main = do
+    withSystemTempDirectory "" $ \stagingDir -> do
+        -- let stagingDir = "C:\\Users\\mwurb\\AppData\\Local\\Temp\\-777f232250ff9e9c"
+        prepareEnvironment stagingDir
+
+        repoDir <- getEnvDefault "APPVEYOR_BUILD_FOLDER" "C:\\dev\\Dataframes"
+        let dataframesSolutionPath = repoDir </> "native_libs" </> "src" </> "DataframeHelper.sln"
+        buildWithMsBuild dataframesSolutionPath
+
+        let packageRoot = stagingDir </> "Dataframes"
+        let packageBinaries = packageRoot </> "native_libs" </> "windows"
+        let builtBinariesDir = repoDir </> "native_libs" </> "src" </> "x64" </> "Release"
+        let packageFile = "Dataframes-Win-x64-v141" <.> "7z"
+        downloadAndUnpack7z packageBaseUrl packageBinaries
+        builtDlls <- glob (builtBinariesDir </> "*.dll")
+        when (null builtDlls) $ error "failed to found built .dll files"
```

Comment:
Very ugly way to exit program

Reply:
What would the pretty way look like here?

Reply:
wrap rest of the program in case or if block (recommend), or use `fail` (I'll prefer using the second one only if there are many ways to fail). This is still not elegant as wrapping code in blocks, but when there are many blocks it may be hard to read.

`error` is rather used in situations that should never happen and you cannot recover if it happened. This is why it is displaying number of line where it occurred. I suggest not using it in production code

Reply:
I'd say `situations that should never happen and you cannot recover if it happened` applies in this case.

Reply:
This is an unexpected, unrecoverable error caused by an external component violating the contract. This can happen when build script reported success but there are no build artifacts. Continuing execution makes no sense (there are no binaries to package) but we also cannot tell what went wrong, as we were told that building binaries succeeded.